origin-js/src/resources/marketplace.js

```diff
@@ -244,52 +252,74 @@ export default class Marketplace {
-    const commissionPerUnit = listing.commissionPerUnit
+    const commissionPerUnit = isMultiUnit
       ? BigNumber(listing.commissionPerUnit.amount)
       : commission
-    const offers = []
-    allOffers.forEach(offer => {
-      const unitsSoldBeforeOffer = listing.unitsTotal - unitsAvailable
-
-      // Validate units purchased against units available.
-      // TODO: Validate units purchased vs. units available at offer creation
-      // time. This will handle the case where an edit of a listing invalidates
-      // a previously valid offer.
-      if (offer.unitsPurchased > unitsAvailable) {
-        return
-      }
-      if (offer.status !== 'withdrawn') {
-        unitsAvailable -= offer.unitsPurchased
-      }
+    let commissionAvailable = new BigNumber(commission) // create new instance
+    let unitsAvailable = listing.unitsTotal
+
+    const offers = allOffers.filter(offer => {
+      const offerCommission = offer.commission && BigNumber(offer.commission.amount)
 
-      // Validate that the offer commission is what we expect. If the amount
-      // of commission for the listing isn't sufficient for this offer, we
-      // require that the offer have whatever commission is available to it.
-      if (commissionPerUnit !== null && commissionPerUnit.isGreaterThan(0)) {
-        const commissionUsed = BigNumber(commissionPerUnit).times(unitsSoldBeforeOffer)
-        const remainingCommission = BigNumber.max(
-          commission.minus(commissionUsed),
+      // required for multi-unit commission and unitsAvailable calculations
+      const onValidOffer = () => {
+        if (offer.status === 'withdrawn')
+          return
+
+        unitsAvailable -= offer.unitsPurchased
+        commissionAvailable = BigNumber.max(
+          commissionAvailable.minus(offerCommission),
           BigNumber(0)
         )
-        const expectedCommission = BigNumber.min(
-          remainingCommission,
-          commissionPerUnit.times(offer.unitsPurchased)
-        )
-        const offerCommission = offer.commission && BigNumber(offer.commission.amount)
-        if (!offerCommission || !offerCommission.isEqualTo(expectedCommission)) {
-          return
+      }
+      /* Validate units purchased against units available.
+       *
+       * TODO: Validate units purchased vs. units available at offer creation
+       * time. This will handle the case where an edit of a listing invalidates
+       * a previously valid offer.
+       * - note: even after the todo implemented there is still an edge case where
+       *   2 users create an offer at the same time.
+       */
+      if (offer.unitsPurchased > unitsAvailable) {
+        return false
+      }
+
+      // listing has no commission
+      if (!isMultiUnit && (commissionPerUnit === null || commissionPerUnit.isEqualTo(0))) {
+        onValidOffer()
+        return true
+      }
+      else if (!isMultiUnit) {
+        if (offerCommission && offerCommission.isEqualTo(commission)){
```

Comment:
is there corner case where the seller increases the commission through a listing update and invalidates previous offers that have the original commission? if so, a TODO would be good here

Reply:
good thinking. Currently a seller can not edit the overall commission (boostLimit) and commission per unit (boost) when editing a listing. Will add still add a comment, just in case we enable that in the future.